core/modules/filter/filter.module

```diff
@@ -1585,6 +1585,62 @@ function filter_dom_serialize_escape_cdata_element($dom_document, $dom_element,
 }
 
 /**
+ * Implements hook_form_FORM_ID_alter().
+ */
+function filter_form_filter_format_editor_image_form_alter(&$form, &$form_state) {
+  // Mark files uploaded on entities as temporary, as they will be updated and
+  // added into the file usage table in filter_entity_insert().
+  $entity = _filter_page_is_entity();
+  if (is_object($entity)) {
+    $form['image']['fid']['#temporary'] = TRUE;
+  }
+}
+
+/**
+ * Determine if the page we are on belongs to any entity.
+ *
+ * @param  $path
+ *   URL or path.
+ *
+ * @return boolean|Entity
+ *   Returns the entity loaded by the menu mapping at this path.
+ *   Returns FALSE if the path does not load an entity.
+ */
+function _filter_page_is_entity() {
+  $path = _filter_referer_path();
+  $router_item = menu_get_item($path);
```

Comment:
I don't think I'd trust this `_filter_referer_path()` function. If we need to determine whether or not the image dialog is on an entity or not, we should pass in a query string that could be checked rather than trying to reverse-engineer if we came from a entity edit form.